UserDetailsClient/UserDetailsClient.Core/Features/LogOn/B2CAuthenticationService.cs

```diff
@@ -0,0 +1,157 @@
+                .WithParentActivityOrWindow(ParentActivityOrWindow)
+                .ExecuteAsync();
+
+            var userContext = UpdateUserInfo(ar);
+
+            return userContext;
+        }
+
+        public async Task<UserContext> EditProfile()
+        {
+            IEnumerable<IAccount> accounts = await PCA.GetAccountsAsync();
+            // KNOWN ISSUE:
+            // User will get prompted 
+            // to pick an IdP again.
+            AuthenticationResult ar = await PCA.AcquireTokenInteractive(B2CConstants.Scopes)
+                .WithAccount(GetAccountByPolicy(accounts, B2CConstants.PolicyEditProfile))
+                .WithPrompt(Prompt.NoPrompt)
+                .WithAuthority(B2CConstants.AuthorityEditProfile)
+                .WithParentActivityOrWindow(ParentActivityOrWindow)
+                .ExecuteAsync();
+
+            var userContext = UpdateUserInfo(ar);
+
+            return userContext;
+        }
+
+        public async Task<UserContext> SignIn()
+        {
+            IEnumerable<IAccount> accounts = await PCA.GetAccountsAsync();
+            AuthenticationResult ar = await PCA.AcquireTokenInteractive(B2CConstants.Scopes)
+                .WithAccount(GetAccountByPolicy(accounts, B2CConstants.PolicySignUpSignIn))
+                .WithParentActivityOrWindow(ParentActivityOrWindow)
+                .ExecuteAsync();
+
+            var newContext = UpdateUserInfo(ar);
+            return newContext;
+        }
+
+        public async Task<UserContext> SignOut()
+        {
+
+            IEnumerable<IAccount> accounts = await PCA.GetAccountsAsync();
+            while (accounts.Any())
+            {
+                await PCA.RemoveAsync(accounts.FirstOrDefault());
+                accounts = await PCA.GetAccountsAsync();
+            }
+            var signedOutContext = new UserContext();
+            signedOutContext.IsLoggedOn = false;
+            return signedOutContext;
+        }
+
+        private IAccount GetAccountByPolicy(IEnumerable<IAccount> accounts, string policy)
+        {
+            foreach (var account in accounts)
+            {
+                string userIdentifier = account.HomeAccountId.ObjectId.Split('.')[0];
+                if (userIdentifier.EndsWith(policy.ToLower())) return account;
+            }
+
+            return null;
+        }
+
+
+        private string Base64UrlDecode(string s)
+        {
+            s = s.Replace('-', '+').Replace('_', '/');
+            s = s.PadRight(s.Length + (4 - s.Length % 4) % 4, '=');
+            var byteArray = Convert.FromBase64String(s);
+            var decoded = Encoding.UTF8.GetString(byteArray, 0, byteArray.Count());
+            return decoded;
+        }
+
+        public UserContext UpdateUserInfo(AuthenticationResult ar)
+        {
+            var newContext = new UserContext();
+            newContext.IsLoggedOn = false;
+            JObject user = ParseIdToken(ar.IdToken);
```

Comment:
The AuthenticationResult should have some of these. Is it important to you to parse the IdToken to get all these extra details?

Reply:
I think this is how the sample was set-up originally. I think it's best to make an issue and update this in a future PR. your thoughts, @bogdan Gavril ?

---
In reply to: [280932304](https://github.com/Azure-Samples/active-directory-b2c-xamarin-native/pull/85#discussion_r280932304) [](ancestors = 280932304)

Reply:
It's fine, we can always change the sample later.

Reply:
i like to get the email out personally.

Reply:
+1

---
In reply to: [280934072](https://github.com/Azure-Samples/active-directory-b2c-xamarin-native/pull/85#discussion_r280934072) [](ancestors = 280934072)